### PR TITLE
substitute_one_step: stop after first variable has been substituted

### PR DESCRIPTION
**Pull request by cre4ture**
This change provides the posibility to only replace the next variable in a given string, returning the position after the variable name. 
This is usefull if the substitution is to be integrated into an existing string parsing statemachine.
E.g. I'm integrating in into a "shell-words" like funktionality needed by the uutils/coreutils-env app.

To avoid that we need to do a fork of your code, we want to bring it into your main branch.
Please tell me what you think.

Kind regards

### CODE REVIEW

**Comment by de-vri-es** [comment]
Hey! Sounds interesting. But maybe we can make it return a tuple of 3 slices instead of a `Option<(String, usize)>`: the leading part of the string before the next variable, the substitute value, and the remainder of the unprocessed string.

**Comment by cre4ture** [comment]
You where right, the interface was a bit mixed up. I did refactorings and I think this fits now more to your idea.
What do you think?

**Comment by de-vri-es** [comment]
I think we still need to prevent returning a `String`, since it will re-allocate every time a change is made now.

But I also wonder, do you still need `substitute_one(...)` if we have a `parse(..) -> Template` function? Or can you achieve your desired end result already if parsing and performing substitution is split?

File: src/lib.rs [hunk 1]
```diff
@@ -418,7 +494,10 @@ mod test {
 	fn substitution_in_default_value() {
 		let mut map: BTreeMap<String, String> = BTreeMap::new();
 		map.insert("name".into(), "world".into());
-		check!(let Ok("Hello cruel world!") = substitute("Hello ${not_name:cruel $name}!", &map).as_deref());
+		assert_eq!(
+			Ok("Hello cruel world!"),
+			substitute("Hello ${not_name:cruel $name}!", &map).as_deref()
+		);
```
Comment (de-vri-es) [inline src/lib.rs:500]: Don't replace these. The `check!()` and `assert!()` macros from [`assert2`](https://docs.rs/assert2) give far better error messages than `assert_eq!()`.

File: src/lib.rs [hunk 2]
```diff
@@ -591,7 +670,7 @@ mod test {
 		let variables: &dyn VariableMap<Value = &String> = &variables;
 
 		let_assert!(Ok(expanded) = substitute("one ${aap}", variables));
-		assert!(expanded == "one noot");
+		assert_eq!(expanded, "one noot");
```
Comment (de-vri-es) [inline src/lib.rs:673]: Same here, don't change this. [`assert2::assert!()`](https://docs.rs/assert2/latest/assert2/macro.assert.html) is better.

File: src/lib.rs [hunk 3]
```diff
@@ -607,4 +686,34 @@ mod test {
...
+	fn test_substitute_one_step_variable_and_escape_sequence() {
+		let mut variables = BTreeMap::new();
+		variables.insert(String::from("NAME"), String::from("subst"));
+
+		let source = r"hello $NAME. Nice\$to meet you $NAME.";
+		let result = substitute_one_step(source, &variables).unwrap().unwrap();
+		assert_eq!(result.0, "subst");
+		assert_eq!(result.1.slice_before_ends, 6);
+		assert_eq!(result.1.slice_after_starts, 11);
+		assert_eq!(result.1.subst_type, SubstitutionType::Variable);
```
Comment (de-vri-es) [inline src/lib.rs:700]:
```suggestion
		assert!(result.0 == "subst");
		assert!(result.1.slice_before_ends == 6);
		assert!(result.1.slice_after_starts == 11);
		assert!(result.1.subst_type == SubstitutionType::Variable);
```

File: src/lib.rs [hunk 4]
```diff
@@ -607,4 +686,34 @@ mod test {
...
 		));
 	}
+
+	#[test]
+	fn test_substitute_one_step_variable_and_escape_sequence() {
+		let mut variables = BTreeMap::new();
+		variables.insert(String::from("NAME"), String::from("subst"));
+
+		let source = r"hello $NAME. Nice\$to meet you $NAME.";
+		let result = substitute_one_step(source, &variables).unwrap().unwrap();
```
Comment (de-vri-es) [inline src/lib.rs:696]:
```suggestion
		let_assert!(Ok(Some(result)) = substitute_one_step(source, &variables));
```

File: src/lib.rs [hunk 5]
```diff
@@ -607,4 +686,34 @@ mod test {
...
+		assert_eq!(result.1.slice_after_starts, 11);
+		assert_eq!(result.1.subst_type, SubstitutionType::Variable);
+
+		let result = substitute_one_step(source.get(result.1.slice_after_starts..).unwrap(), &variables)
+			.unwrap()
+			.unwrap();
+		assert_eq!(result.0, "$");
+		assert_eq!(result.1.slice_before_ends, 6);
+		assert_eq!(result.1.slice_after_starts, 8);
+		assert_eq!(result.1.subst_type, SubstitutionType::UnescapeOne);
```
Comment (de-vri-es) [inline src/lib.rs:708]:
```suggestion
		let_assert!(Ok(Some(result)) = substitute_one_step(source.get(result.1.slice_after_starts..).unwrap(), &variables));
		assert!(result.0 == "$");
		assert!(result.1.slice_before_ends == 6);
		assert!(result.1.slice_after_starts == 8);
		assert!(result.1.subst_type == SubstitutionType::UnescapeOne);
```

File: src/lib.rs [hunk 6]
```diff
@@ -607,4 +686,34 @@ mod test {
...
+	}
+
+	#[test]
+	fn test_substitute_one_step_no_substitution() {
+		let variables: BTreeMap<String, String> = BTreeMap::new();
+
+		let source = r"hello world";
+		let result = substitute_one_step(source, &variables).unwrap();
+		assert!(result.is_none());
```
Comment (de-vri-es) [inline src/lib.rs:717]:
```suggestion
		assert!(let Ok(None) = substitute_one_step(source, &variables));
```

**Review by de-vri-es — COMMENTED** [review]
For unit tests, we use [`assert2`](https://docs.rs/assert2) :)

**Comment by cre4ture** [comment]
> But I also wonder, do you still need `substitute_one(...)` if we have a `parse(..) -> Template` function? Or can you achieve your desired end result already if parsing and performing substitution is split?

I would still need a funktion like `parse_one_template_part` as I only trigger the variable substitution when  I detected that there is a $. The part of the input after that variable is again parsed with my own parser.

But we can go for the template based approach anyway if you like. Just let me know what you thing about #14

**Comment by de-vri-es** [comment]
> I would still need a funktion like `parse_one_template_part` as I only trigger the variable substitution when I detected that there is a $. The part of the input after that variable is again parsed with my own parser.

I think I'm not entirely seeing your use case clearly. It feels somewhat odd to perform only partial variable substitution. Wouldn't it make more sense to run your own parser, and then run the substitution on the relevant parts that your parser indicates?

Or maybe the opposite makes sense: parse a full template first, and then run your parser on the non-variable parts of the template?

Do you have a link to how you're using this maybe? Or otherwise could you explain with some pseudo-code how this is supposed to integrate with your parser?



> But we can go for the template based approach anyway if you like. Just let me know what you thing about #14

I will! Probably not today though, as I'm on holiday :p So I have only limited time for reviews. But the comments about the unit tests apply to both PRs. If you have the time you could already process those.

**Comment by cre4ture** [comment]
This pull request contains (temporarily) the modified subst such that it works. If you like you can have a look at it:
https://github.com/uutils/coreutils/pull/5801

I think the main problems I encountered when I tried to use the unmodified subst where these:
- When running the subst befor or after my parser: Unescaping was done twice such that some tests where this is relevant where failing. E.g.`\$NAME` should not be substituted, but was because my parser needs to deal and remove the escape char for other cases anyway.
- When running subst only on the variables: I would need to implement the detection of the end of the variable name myself. Only this way I can forward the exact relevant part.

With the "one-step" way, I just need to stop at an unescaped $ and call subst_one_step which tells me about the end of the variable name and the substituted value.

In the beginning, when I didn't know the subst source code that well, for me the easiest way to achieve my goal was to provide access to this single/one-substitution step. Now, as I know your code pretty well, I detected that maybe the only function that I need from subst is the `parse_variable`. Making that function public woult probably be enought.

**Comment by de-vri-es** [comment]
Hmm, interesting. I still like the template approach, but I also like to unblock progress for `uutils`. I'll have a look at that PR.

**Pull request closed without merging**